### PR TITLE
Density matrix sampling for terminal measurements

### LINKED ISSUE #1381: Density matrix simulator should be optimized for terminal measurements
similar to other simulators.

### PR DESCRIPTION
**Pull request by balopat**
Fixes #1381.

### CODE REVIEW

File: cirq/sim/density_matrix_simulator.py [hunk 1]
```diff
@@ -123,12 +123,31 @@ def _run(self,
...
-        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+        resolved_circuit = protocols.resolve_parameters(circuit,
+                                                        param_resolver)
+
+        if circuit.are_all_measurements_terminal():
+            return self._run_sweep_sample(resolved_circuit, repetitions)
+        else:
+            return self._run_sweep_repeat(resolved_circuit, repetitions)
+
+    def _run_sweep_sample(self, circuit, repetitions):
```
Comment (Strilanc) [inline cirq/sim/density_matrix_simulator.py]: Add type annotations

File: cirq/sim/density_matrix_simulator.py [hunk 2]
```diff
@@ -123,12 +123,31 @@ def _run(self,
...
+                qubit_order=ops.QubitOrder.DEFAULT,
+                initial_state=0,
+                perform_measurements=False):
+            pass
+        measurement_ops = [op for _, op, _ in
+                           circuit.findall_operations_with_gate_type(
+                               ops.MeasurementGate)]
+        return step_result.sample_measurement_ops(measurement_ops, repetitions)
+
+    def _run_sweep_repeat(self, circuit, repetitions):
```
Comment (Strilanc) [inline cirq/sim/density_matrix_simulator.py]: Add type annotations

File: cirq/sim/density_matrix_simulator_test.py [hunk 3]
```diff
@@ -115,10 +115,30 @@ def test_run_channel(dtype):
...
+class CountingSimulator(cirq.DensityMatrixSimulator):
+    """
+    CountingSimulator is a test class that decorates the _base_iterator method
+    so that we can count the number of circuit runs during a test case
+    """
+    def __init__(self, dtype=np.complex64):
+        super().__init__(dtype)
+        self.count = 0
+
+    def _base_iterator(
```
Comment (Strilanc) [inline cirq/sim/density_matrix_simulator_test.py]: I think that this kind of interception is more typically done via mocking.

File: cirq/sim/sparse_simulator_test.py [hunk 4]
```diff
@@ -66,10 +67,30 @@ def test_run_bit_flips(dtype):
...
+        self.count = 0
+
+    def _base_iterator(
+            self,
+            circuit: cirq.Circuit,
+            qubit_order: cirq.ops.QubitOrderOrList,
+            initial_state: Union[int, np.ndarray],
+            perform_measurements: bool = True,
+    ):
+        self.count += 1
```
Comment (Strilanc) [inline cirq/sim/sparse_simulator_test.py]: Use mocks instead.

(Note: python 2 and python 3 get their mocks from different places, so we have `cirq.testing.mock` to hook up the right one.)

File: cirq/sim/density_matrix_simulator.py (hunk 2 above)
Comment (balopat) [reply cirq/sim/density_matrix_simulator.py]: ack

File: cirq/sim/density_matrix_simulator_test.py (hunk 3 above)
Comment (balopat) [reply cirq/sim/density_matrix_simulator_test.py]: Of course! Thanks.
it was one of those late night moments where I looked and couldn't find it in unittest.mock and then I just implemented myself.
yup, it's much simpler with mocking library. change is coming...

File: cirq/sim/sparse_simulator_test.py (hunk 4 above)
Comment (balopat) [reply cirq/sim/sparse_simulator_test.py]: ack - thanks for pointing me at cirq.testing.mock.

**Comment by balopat** [comment]
@Strilanc ready for another look.

**Pull request merged**